netZooPy/panda/panda.py

```diff
@@ -420,32 +420,68 @@ def update_diagonal(diagonal_matrix, num, alpha, step):
+            return a_matrix
+
+        def gupdate_diagonal(diagonal_matrix, num, alpha, step):
+            '''Update diagonal.'''
+            cp.fill_diagonal(diagonal_matrix, cp.nan)
+            diagonal_std = cp.nanstd(diagonal_matrix, 1)
+            diagonal_fill = diagonal_std * num * math.exp(2 * alpha * step)
+            cp.fill_diagonal(diagonal_matrix, diagonal_fill)
+
         panda_loop_time = time.time()
         num_tfs, num_genes = motif_matrix.shape
         step = 0
         hamming = 1
         alpha = 0.1
+        
         while hamming > 0.001:
             # Update motif_matrix
-            W = 0.5 * (t_function(ppi_matrix, motif_matrix) + t_function(motif_matrix, correlation_matrix))  # W = (R + A) / 2
-            hamming = np.abs(motif_matrix - W).mean()
-            motif_matrix *= (1 - alpha)
-            motif_matrix += (alpha * W)
-
-            if hamming > 0.001:
+            if computing=='gpu':
+                import cupy as cp
+                ppi_matrix=cp.array(ppi_matrix)
+                motif_matrix=cp.array(motif_matrix)
+                correlation_matrix=cp.array(correlation_matrix)
+                W = 0.5 * (gt_function(ppi_matrix, motif_matrix) + gt_function(motif_matrix, correlation_matrix))  # W = (R + A) / 2
+                hamming = cp.abs(motif_matrix - W).mean()
+                motif_matrix *= (1 - alpha)
+                motif_matrix += (alpha * W)
+                ppi = gt_function(motif_matrix)  # t_func(X, X.T)
+                motif = gt_function(motif_matrix.T)
                 # Update ppi_matrix
+                gupdate_diagonal(ppi, num_tfs, alpha, step)
+                ppi_matrix *= (1 - alpha)
+                ppi_matrix += (alpha * ppi)
+                # Update correlation_matrix
+                gupdate_diagonal(motif, num_genes, alpha, step)
+                correlation_matrix *= (1 - alpha)
+                correlation_matrix += (alpha * motif)
+            elif computing=='cpu':
+                W = 0.5 * (t_function(ppi_matrix, motif_matrix) + t_function(motif_matrix, correlation_matrix))  # W = (R + A) / 2
+                hamming = np.abs(motif_matrix - W).mean()
+                motif_matrix *= (1 - alpha)
+                motif_matrix += (alpha * W)
                 ppi = t_function(motif_matrix)  # t_func(X, X.T)
+                motif = t_function(motif_matrix.T)
+                # Update ppi_matrix
                 update_diagonal(ppi, num_tfs, alpha, step)
                 ppi_matrix *= (1 - alpha)
                 ppi_matrix += (alpha * ppi)
-
                 # Update correlation_matrix
-                motif = t_function(motif_matrix.T)  # t_func(X.T, X)
                 update_diagonal(motif, num_genes, alpha, step)
                 correlation_matrix *= (1 - alpha)
                 correlation_matrix += (alpha * motif)
 
-                del W, ppi, motif  # release memory for next step
+            del W, ppi, motif  # release memory for next step
```

Comment:
the `del` should have the same indentation level as `correlation_matrix`

Reply:
but then I need it in both if and else?

Reply:
i get it